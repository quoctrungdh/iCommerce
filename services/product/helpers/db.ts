import mongoose from 'mongoose'
import { MONGO_SERVER, MONGO_DB } from '../constants';

export async function initDatabase() {
  if(!MONGO_SERVER) {
    throw new Error("Mongo Server is not found")
  }

  if(!MONGO_DB) {
    throw new Error("Mongo DB is not found")
  }

  console.log({ MONGO_SERVER, MONGO_DB })

  mongoose.Promise = global.Promise;

  await mongoose.connect(MONGO_SERVER, {
    useNewUrlParser: true,
    useUnifiedTopology: true,
    useCreateIndex: true,
    dbName: MONGO_DB
  })

  // Get the default connection
  const db = mongoose.connection;
  

  // Bind connection to error event (to get notification of connection errors)
  db.on('error', console.error.bind(console, 'MongoDB connection error:'));
}
