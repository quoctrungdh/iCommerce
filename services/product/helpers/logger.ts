import winston from 'winston';
import fluentLogger from 'fluent-logger'
import { FLUENTD_HOST } from '../constants'


function loggerHanlder() {
  const config = {
    host: FLUENTD_HOST,
    port: 24224,
    timeout: 3.0,
    requireAckResponse: true // Add this option to wait response from Fluentd certainly
  };
  
  console.log("config", JSON.stringify(config))
  let fluent: any
  try {
    const fluentTransport = fluentLogger.support.winstonTransport()
    fluent = new fluentTransport('product.service', config);
    
  } catch(error) {
    console.error(error)
  }

  const logger = winston.createLogger({
    transports: [fluent, new (winston.transports.Console)()]
  });
  
  logger.on('finish', () => {
    console.log("finish")
    fluent.sender.end("end", {}, () => {})
  });

  return logger
}

export default loggerHanlder()