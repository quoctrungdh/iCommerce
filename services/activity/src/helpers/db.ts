import mongoose from 'mongoose';
import { MONGO_SERVER, MONGO_DB } from '../constants';
import logger from '../helpers/logger';

export default async function initDatabase() {
  if (!MONGO_SERVER) {
    throw new Error('Mongo Server is not found');
  }

  if (!MONGO_DB) {
    throw new Error('Mongo DB is not found');
  }

  mongoose.Promise = global.Promise;

  const readyStatus = [
    'disconnected',
    'connected',
    'connecting',
    'disconnecting',
  ];

  const dbClient = await mongoose.connect(`${MONGO_SERVER}/${MONGO_DB}`, {
    useNewUrlParser: true,
    useUnifiedTopology: true,
    useCreateIndex: true,
  });

  // Get the default connection
  // const db = dbClient.connection;
  // logger.info('readyState', readyStatus[db.readyState]);

  // db.once('open', (res) => {
  //   logger.info('db connection open', JSON.stringify(res));
  // });

  // // Bind connection to error event (to get notification of connection errors)
  // db.on('error', logger.error);

  return dbClient;
}
