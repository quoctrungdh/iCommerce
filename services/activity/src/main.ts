import Koa from 'koa';
import Router from 'koa-router';
import koaLogger from 'koa-logger';
import json from 'koa-json';
import bodyParser from 'koa-bodyparser';
import { PORT, IS_DEV, IS_TEST } from './constants';
import initDatabase from './helpers/db';
import errorHandler from './middlewares/errorHandler';
import {
  getActivities,
  getActivity,
  createActivity,
  updateActivity,
  deleteActivity,
  ping,
} from './controllers/activity.controller';
import logger from './helpers/logger';
import { serializeError } from 'serialize-error';
import NatsHandler from './events/natsHandler';

const app = new Koa();
const router = new Router();
const nc = new NatsHandler();

router
  .get('/ping', ping)
  .get('/', getActivities)
  .get('/:id', getActivity)
  .post('/', createActivity)
  .put('/', updateActivity)
  .delete('/', deleteActivity);

// Middlewares
app.use(errorHandler);
app.use(json());
app.use(bodyParser());
if (IS_DEV) {
  app.use(koaLogger());
}
app.use(nc.natsMiddleware);

// Routes
app.use(router.routes()).use(router.allowedMethods());

/* centralized error handling: */
app.on('error', (err, ctx) => {
  logger.error(JSON.stringify(serializeError(err)));
});

const server = app.listen(PORT, () => {
  initDatabase();
  if (!IS_TEST) {
    logger.info('Koa started', PORT);
  }
});

export default server;
