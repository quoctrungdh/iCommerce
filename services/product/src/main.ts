import Koa from 'koa';
import Router from 'koa-router';
import koaLogger from 'koa-logger';
import json from 'koa-json';
import bodyParser from 'koa-bodyparser';
import { PORT, IS_DEV, IS_TEST } from './constants';
import initDatabase from './helpers/db';
import {
  getProduct,
  getProducts,
  createProduct,
  updateProduct,
  deleteProduct,
  ping,
} from './controllers/product.controller';
import errorHandler from './middlewares/errorHandler';
import logger from './helpers/logger';
import { serializeError } from 'serialize-error';
import NatsHandler from './events/natsHandler';

const app = new Koa();
const router = new Router();
const nc = new NatsHandler();

router
  .get('/ping', ping)
  .get('/', getProducts)
  .get('/:id', getProduct)
  .post('/', createProduct)
  .put('/', updateProduct)
  .delete('/', deleteProduct);

// Middlewares
app.use(errorHandler);
app.use(json());
app.use(bodyParser());
if (IS_DEV) {
  app.use(koaLogger());
}
app.use(nc.natsMiddleware);

// Routes
app.use(router.routes()).use(router.allowedMethods());

/* centralized error handling: */
app.on('error', (err, ctx) => {
  logger.error(JSON.stringify(serializeError(err)));
});

const server = app.listen(PORT, () => {
  initDatabase();
  if (!IS_TEST) {
    logger.info('Koa started', PORT);
  }
});

export default server;
