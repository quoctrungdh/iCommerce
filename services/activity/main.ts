import Koa from "koa";
import Router from "koa-router";
import logger from "koa-logger";
import json from "koa-json";
import { PORT } from './constants'
import { initDatabase } from "./helpers/db";
import errorHandler from "./middlewares/errorHandler";
import { getActivities, getActivity, createActivity, updateActivity, deleteActivity, ping } from "./controllers/activity.controller";
import natsMiddleware from './middlewares/nats'

const app = new Koa();
const router = new Router();

router.get("/ping", ping);
router.get("/", getActivities);
router.get("/:id", getActivity);
router.post("/", createActivity);
router.put("/", updateActivity);
router.delete("/", deleteActivity);

// Middlewares
app.use(errorHandler);
app.use(json());
app.use(logger());
app.use(natsMiddleware)

// Routes
app.use(router.routes()).use(router.allowedMethods());

/* centralized error handling: */
app.on('error', (err, ctx) => {
  console.log("onerror", err)
  /* 
   *   console.log error
   *   write error to log file
   *   save error and request information to database if ctx.request match condition
   *   ...
  */
});

app.listen(PORT, () => {
  initDatabase()
  console.log("Koa started", PORT);
});