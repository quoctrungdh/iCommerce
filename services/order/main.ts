import Koa from "koa";
import Router from "koa-router";
import koaLogger from "koa-logger";
import json from "koa-json";
import bodyParser from 'koa-bodyparser';
import { PORT, IS_DEV } from './constants'
import { initDatabase } from "./helpers/db";
import errorHandler from "./middlewares/errorHandler";
import { ping, getOrders, getOrder, createOrder, updateOrder, deleteOrder } from "./controllers/order.controller";
import logger from './helpers/logger'
import { serializeError } from 'serialize-error';
import NatsHandler from "./events/natsHandler";

const app = new Koa();
const router = new Router();
const nc = new NatsHandler()

router.get("/ping", ping);
router.get("/", getOrders);
router.get("/:id", getOrder);
router.post("/", createOrder);
router.put("/", updateOrder);
router.delete("/", deleteOrder);

// Middlewares
app.use(errorHandler);
app.use(json());
app.use(bodyParser());
if(IS_DEV) {
  app.use(koaLogger());
}
app.use(nc.natsMiddleware)

// Routes
app.use(router.routes()).use(router.allowedMethods());

/* centralized error handling: */
app.on('error', (err, ctx) => {
  logger.error(JSON.stringify(serializeError(err)))
});

app.listen(PORT, () => {
  initDatabase()
  console.log("Koa started", PORT);
});