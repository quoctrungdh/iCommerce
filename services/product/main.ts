import Koa from "koa";
import Router from "koa-router";
import logger from "koa-logger";
import json from "koa-json";
import { PORT } from './constants'
import { initDatabase } from "./helpers/db";
import { getProduct, getProducts, createProduct, updateProduct, deleteProduct } from "./controllers/product.controller";
import errorHandler from "./middlewares/errorHandler";

const app = new Koa();
const router = new Router();

router.get("/", getProducts);
router.get("/:id", getProduct);
router.post("/", createProduct);
router.put("/", updateProduct);
router.delete("/", deleteProduct);

// Middlewares
app.use(errorHandler);
app.use(json());
app.use(logger());

// Routes
app.use(router.routes()).use(router.allowedMethods());

/* centralized error handling: */
app.on('error', (err, ctx) => {
  console.log("onerror", err)
  /* 
   *   console.log error
   *   write error to log file
   *   save error and request information to database if ctx.request match condition
   *   ...
  */
});

app.listen(PORT, () => {
  initDatabase()
  console.log("Koa started", PORT);
});