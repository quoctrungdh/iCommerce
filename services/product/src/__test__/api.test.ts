import mockingoose from 'mockingoose';
import '../helpers/sinon-mongoose';
import productModel from '../models/product.model';
import supertest from 'supertest';
import main from '../main';

jest.mock('ts-nats', () => ({
  connect: () => ({
    on: jest.fn(),
    publish: jest.fn(),
  }),
  Payload: {
    JSON: 'json',
  },
}));
jest.mock('winston', () => ({
  format: {
    colorize: jest.fn(),
    combine: jest.fn(),
    label: jest.fn(),
    timestamp: jest.fn(),
    printf: jest.fn(),
    errors: jest.fn(),
  },
  createLogger: jest.fn().mockReturnValue({
    info: jest.fn(),
    error: jest.fn(),
    debug: jest.fn(),
    log: jest.fn(),
    on: jest.fn(),
  }),
  transports: {
    Console: jest.fn(),
  },
}));
jest.mock('fluent-logger', () => ({
  support: {
    winstonTransport: () => jest.fn(),
  },
}));

const _data = [
  {
    _id: '507f191e810c19729de860ea',
    name: 'Macbook Pro 2020',
    price: 20000,
    brand: 'Apple',
    color: 'grey',
    createdAt: '2020-08-22T08:47:38.094Z',
    updatedAt: '2020-08-22T08:47:03.693Z',
  },
  {
    _id: '5f40da4cca1e00c3cd23f1a0',
    name: 'ÁO POLO NAM',
    price: 30000,
    brand: 'OOOP',
    color: 'red',
    createdAt: '2020-08-22T08:47:38.094Z',
    updatedAt: '2020-08-22T08:47:03.693Z',
  },
  {
    _id: '5f40da5f1fc695c40887b8c1',
    name: 'Samsung Galaxy New',
    price: 1000000,
    brand: 'Samsung',
    color: 'navy',
    createdAt: '2020-08-22T08:47:38.094Z',
    updatedAt: '2020-08-22T08:47:03.693Z',
  },
];

describe('API test', () => {
  afterAll(() => {
    main.close();
  });

  test('GET /list', async () => {
    mockingoose(productModel).toReturn(_data, 'find');
    productModel.find().then(doc => {
      expect(JSON.parse(JSON.stringify(doc))).toMatchObject(_data);
    });
    const response = await supertest(main).get('/');
    expect(response.status).toEqual(200);
    expect(response.text).toMatchSnapshot();
  });

  test('GET / empty data', async () => {
    mockingoose(productModel).toReturn([], 'find');
    productModel.find().then(doc => {
      expect(JSON.parse(JSON.stringify(doc))).toMatchObject([]);
    });
    const response = await supertest(main).get('/');
    expect(response.status).toEqual(200);
    expect(response.body).toEqual({ data: [], meta: { page: 0, pageSize: 0 } });
  });

  test('GET /:id', async () => {
    mockingoose(productModel).toReturn(_data[0], 'findOne');
    productModel.findOne().then(doc => {
      expect(JSON.parse(JSON.stringify(doc))).toMatchObject(_data[0]);
    });
    const response = await supertest(main).get('/507f191e810c19729de860ea');
    expect(response.status).toEqual(200);
    expect(response.body).toEqual(_data[0]);
  });

  test('GET /:id - invalid id', async () => {
    mockingoose(productModel).toReturn(null, 'findOne');
    productModel.findOne().then(doc => {
      expect(JSON.parse(JSON.stringify(doc))).toBeNull();
    });
    const response = await supertest(main).get('/invalid-id-fake');
    expect(response.status).toEqual(404);
  });

  test('POST /', async () => {
    mockingoose(productModel).toReturn(null, 'save');
    productModel.create().then(doc => {
      expect(doc).toBeUndefined();
    });
  });

  test('PUT /:id', async () => {
    const output = {
      _id: '5f40da5f1fc695c40887b8c1',
      name: 'Samsung Galaxy New',
      price: 5000000,
      brand: 'Samsung',
      color: 'navy',
      inventory: 5,
      createdAt: new Date('2020-08-22T08:47:38.094Z'),
      updatedAt: new Date('2020-08-22T08:47:03.693Z'),
    };
    mockingoose(productModel).toReturn(output, 'findOneAndUpdate');
    productModel.findOneAndUpdate(
      { _id: '5f40da5f1fc695c40887b8c1' },
      { price: 5000000 })
    .exec().then(doc => {
      expect(doc).toMatchSnapshot();
    });
  });

  test('DELETE /:id', async () => {
    const output = {
      _id: '5f40da5f1fc695c40887b8c1',
      createdAt: new Date('2020-08-22T09:23:57.318Z'),
      updatedAt: new Date('2020-08-22T09:23:57.318Z'),
    };
    mockingoose(productModel).toReturn(output, 'deleteOne');
    productModel.deleteOne(
      { _id: '5f40da5f1fc695c40887b8c1' },
    ).exec().then(doc => {
      expect(doc).toMatchSnapshot();
    });
  });
});
