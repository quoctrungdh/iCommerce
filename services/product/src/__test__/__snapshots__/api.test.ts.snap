// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`API test DELETE /:id 1`] = `
Object {
  "_id": "5f40da5f1fc695c40887b8c1",
  "createdAt": 2020-08-22T09:23:57.318Z,
  "updatedAt": 2020-08-22T09:23:57.318Z,
}
`;

exports[`API test GET /list 1`] = `
"{
  \\"data\\": [
    {
      \\"createdAt\\": \\"2020-08-22T08:47:38.094Z\\",
      \\"updatedAt\\": \\"2020-08-22T08:47:03.693Z\\",
      \\"_id\\": \\"507f191e810c19729de860ea\\",
      \\"name\\": \\"Macbook Pro 2020\\",
      \\"price\\": 20000,
      \\"brand\\": \\"Apple\\",
      \\"color\\": \\"grey\\"
    },
    {
      \\"createdAt\\": \\"2020-08-22T08:47:38.094Z\\",
      \\"updatedAt\\": \\"2020-08-22T08:47:03.693Z\\",
      \\"_id\\": \\"5f40da4cca1e00c3cd23f1a0\\",
      \\"name\\": \\"√ÅO POLO NAM\\",
      \\"price\\": 30000,
      \\"brand\\": \\"OOOP\\",
      \\"color\\": \\"red\\"
    },
    {
      \\"createdAt\\": \\"2020-08-22T08:47:38.094Z\\",
      \\"updatedAt\\": \\"2020-08-22T08:47:03.693Z\\",
      \\"_id\\": \\"5f40da5f1fc695c40887b8c1\\",
      \\"name\\": \\"Samsung Galaxy New\\",
      \\"price\\": 1000000,
      \\"brand\\": \\"Samsung\\",
      \\"color\\": \\"navy\\"
    }
  ],
  \\"meta\\": {
    \\"page\\": 0,
    \\"pageSize\\": 3
  }
}"
`;

exports[`API test PUT /:id 1`] = `
Object {
  "_id": "5f40da5f1fc695c40887b8c1",
  "brand": "Samsung",
  "color": "navy",
  "createdAt": 2020-08-22T08:47:38.094Z,
  "name": "Samsung Galaxy New",
  "price": 5000000,
  "updatedAt": 2020-08-22T08:47:03.693Z,
}
`;
